### PR TITLE
Updating ggnn model implementation

### PR DESCRIPTION
**Pull request by Zacharias030**
Failing 2/8 tests from `ggnn_tests.py`. Unclear why.

I commented the code for you to share my recent learnings.

Some bugs from the old code were indeed fixed, I believe. Those are marked with a note ` # NB(CHRIS)`.

Minor things maybe need to be lifted into the experimental framework around it that we better do together. (If we want to set certain config values).

### CODE REVIEW

**Comment by ChrisCummins** [comment]
Thanks Zach! I'll take a look. BTW I overhauled my pre-commit formatters into a new package:

```
$ ./tools/format/install.sh
$ format --install_pre_commit_hook
```

For python code there shouldn't be any significant changes, but you do get a shiny new "signed off" by footer on commits :D

**Comment by ChrisCummins** [comment]
I'll merge this after CI passes.

**Comment by ChrisCummins** [comment]
I added the extra tests for different flag combinations and found a few minor issues. Want to take a look?

File: deeplearning/ml4pl/models/ggnn/ggnn.py [hunk 1]
```diff
@@ -210,12 +221,18 @@ def __init__(self, *args, **kwargs):
...
+      self.model.dev,
+      torch.get_default_dtype(),
+    )
 
     if DEBUG:
       for submodule in self.model.modules():
         submodule.register_forward_hook(nan_hook)
 
-    self.model.to(self.dev)
+    self.model.to(self.model.dev)
```
Comment (Zacharias030) [inline deeplearning/ml4pl/models/ggnn/ggnn.py]: this line shouldn't exist. it's old code and superceded by the model moving itself to model.dev before instanciating an optimizer!

File: deeplearning/ml4pl/models/ggnn/ggnn_config.py [hunk 2]
```diff
@@ -23,14 +23,24 @@
...
     self.lr: float = FLAGS.learning_rate
-    self.clip_grad_norm: bool = FLAGS.clamp_gradient_norm  # use 6.0 as default! Set to 0.0 for no clipping.
+    self.clip_grad_norm: bool = FLAGS.clamp_gradient_norm  # use 6.0 as default when clipping! Set to 0.0 for no clipping.
 
     self.vocab_size: int = 8568
     self.inst2vec_embeddings = FLAGS.inst2vec_embeddings
     self.emb_size: int = 200
+
+    # TODO This should be turned off on devmap!
     self.use_selector_embeddings: bool = True
```
Comment (Zacharias030) [inline deeplearning/ml4pl/models/ggnn/ggnn_config.py]: This should be lifted into a flag to set to false on devmap. Or something more hacky.

File: deeplearning/ml4pl/models/ggnn/ggnn_modules.py [hunk 3]
```diff
@@ -41,67 +42,102 @@ class GGNNModel(nn.Module):
...
     # carries two momentum params per trainable model parameter.
+
+    # move model to device before making optimizer!
+    self.dev = (
+      torch.device("cuda")
+      if gpu_scheduler.LockExclusiveProcessGpuAccess()
+      else torch.device("cpu")
+    )
+
+    self.to(self.dev)
```
Comment (Zacharias030) [inline deeplearning/ml4pl/models/ggnn/ggnn_modules.py]: here it moves ...

File: deeplearning/ml4pl/models/ggnn/ggnn_modules.py [hunk 4]
```diff
@@ -41,67 +42,102 @@ class GGNNModel(nn.Module):
...
+    metrics_tuple = self.metrics(logits, labels)
 
     outputs = (
-      (prediction,) + metrics_tuple + (graph_features,) + tuple(unroll_stats)
+      (logits,) + metrics_tuple + (graphwise_readout,) + tuple(unroll_stats)
     )
 
     return outputs
 
+  def num_parameters(self) -> int:
```
Comment (Zacharias030) [inline deeplearning/ml4pl/models/ggnn/ggnn_modules.py]: have you noticed? :)

File: deeplearning/ml4pl/models/ggnn/ggnn_modules.py [hunk 5]
```diff
@@ -211,10 +258,18 @@ def __init__(self, config, pretrained_embeddings=None):
...
     else:
       raise NotImplementedError(config.inst2vec_embeddings)
 
-    if config.use_selector_embeddings:
+    if (
+      hasattr(config, "use_selector_embeddings")
+      and config.use_selector_embeddings
+    ):
       selector_init = torch.tensor(
         # TODO(github.com/ChrisCummins/ProGraML/issues/27): x50 is maybe a
```
Comment (Zacharias030) [inline deeplearning/ml4pl/models/ggnn/ggnn_modules.py]: I wonder if this is an issue, but please keep it in mind when restarting unrolling experiments.

File: deeplearning/ml4pl/models/ggnn/ggnn.py (hunk 1 above)
Comment (ChrisCummins) [reply deeplearning/ml4pl/models/ggnn/ggnn.py]: Should the entire line be removed, or just rever the diff to `self.model.to(self.dev)`?

File: deeplearning/ml4pl/models/ggnn/ggnn_modules.py (hunk 4 above)
Comment (ChrisCummins) [reply deeplearning/ml4pl/models/ggnn/ggnn_modules.py]: <3

File: deeplearning/ml4pl/models/ggnn/ggnn_config.py (hunk 2 above)
Comment (ChrisCummins) [reply deeplearning/ml4pl/models/ggnn/ggnn_config.py]: Done 491432b339cbeb46ab23cd885373ba0b2adb7a2c

File: deeplearning/ml4pl/models/ggnn/ggnn_modules.py (hunk 3 above)
Comment (ChrisCummins) [reply deeplearning/ml4pl/models/ggnn/ggnn_modules.py]: Oh I see now. Done 694b7b7268d18fc5811b9b121cc3e443e890a627

File: deeplearning/ml4pl/models/ggnn/ggnn_modules.py (hunk 5 above)
Comment (ChrisCummins) [reply deeplearning/ml4pl/models/ggnn/ggnn_modules.py]: Wiill do!

**Comment by ChrisCummins** [comment]
@Zacharias030 one last change before the merge bd5122c0651e3c66e16f2ecf37c29f2a30ed832d

Should be just cosmetic, inspired by github.com/ChrisCummins/phd/issues/84

All tests pass!

```
$ bazel test //deeplearning/ml4pl/models/ggnn/...
//deeplearning/ml4pl/models/ggnn/test:load_restore_test                  PASSED in 115.5s
  Stats over 8 runs: max = 115.5s, min = 90.3s, avg = 102.3s, dev = 10.2s
//deeplearning/ml4pl/models/ggnn/test:graph_classification_test          PASSED in 255.0s
  Stats over 16 runs: max = 255.0s, min = 186.3s, avg = 221.4s, dev = 18.2s
//deeplearning/ml4pl/models/ggnn/test:node_classification_test           PASSED in 322.9s
  Stats over 16 runs: max = 322.9s, min = 181.6s, avg = 236.4s, dev = 37.0s

Executed 3 out of 3 tests: 3 tests pass.
```

**Comment by Zacharias030** [comment]
Really good! let's merge?

**Comment by ChrisCummins** [comment]
🎉

**Pull request merged**